contracts/src/ModuleRegistry.sol

```diff
@@ -127,18 +131,55 @@ contract ModuleRegistry is OwnableUpgradeable {
     bytes[] memory validationPayloads,
     uint256 value
   ) public {
-    // If no modules provided, bypass module validation
+    // If no module provided, bypass module validation
     if (modulesAddresses.length == 0) return;
     // Each module involved must have a corresponding item from the validation payload
     if (modulesAddresses.length != validationPayloads.length) revert ModuleValidationPayloadMismatch();
 
-    // For each module check if it is registered and call run method
+    // For each module, check if it is registered and call its run method
     for (uint32 i = 0; i < modulesAddresses.length; i = uncheckedInc32(i)) {
       if (!isRegistered(modulesAddresses[i])) revert ModuleNotRegistered();
       AbstractModule(modulesAddresses[i]).run(attestationPayload, validationPayloads[i], tx.origin, value);
     }
   }
 
+  /**
+   * @notice Executes the V2 run method for all given Modules that are registered
+   * @param modulesAddresses the addresses of the registered modules
+   * @param attestationPayload the payload to attest
+   * @param validationPayloads the payloads to check for each module (one payload per module)
+   * @param value the value (ETH) optionally passed in the attesting transaction
+   * @param initialCaller the address of the initial caller (transaction sender)
+   * @param attester the address defined by the Portal as the attester for this payload
+   * @dev check if modules are registered and execute the V2 run method for each module
+   */
+  function runModulesV2(
+    address[] memory modulesAddresses,
+    AttestationPayload memory attestationPayload,
+    bytes[] memory validationPayloads,
+    uint256 value,
+    address initialCaller,
+    address attester
+  ) public {
+    // If no module provided, bypass module validation
+    if (modulesAddresses.length == 0) return;
+    // Each module involved must have a corresponding item from the validation payload
+    if (modulesAddresses.length != validationPayloads.length) revert ModuleValidationPayloadMismatch();
+
+    // For each module, check if it is registered and call its run method
+    for (uint32 i = 0; i < modulesAddresses.length; i = uncheckedInc32(i)) {
+      if (!isRegistered(modulesAddresses[i])) revert ModuleNotRegistered();
+      AbstractModuleV2(modulesAddresses[i]).run(
+        attestationPayload,
+        validationPayloads[i],
+        initialCaller,
+        value,
+        attester,
+        msg.sender
```

Comment:
Love it! Also love that you included the `initialCaller`. That way we can tell what was up the call chain. 👏